### DISCUSSION TITLE
MAX1704x help needed

### DISCUSSION THREAD

**Discussion by Max-LAB** [UncleRus/esp-idf-lib#585] [Q&A] [answered]
Hi all,

I am trying for my project to use `max1704x.h` library, and this example in here https://github.com/UncleRus/esp-idf-lib/blob/master/examples/max1704x/default/main/main.c. The Problem is that I am using [Adafruit ESP32 S3 Feather board](https://learn.adafruit.com/adafruit-esp32-s3-feather/power-management) that have already integrated MAX17048 chip. 

After an initialisation, board just freez

```c
    ESP_ERROR_CHECK(i2cdev_init());
    esp_err_t r;
    max1704x_t dev = {0 };
    max1704x_config_t config = { 0 };
    max1704x_status_t status = { 0 };
    uint16_t version = 0;
    float voltage = 0;
    float soc_percent = 0;
    float rate_change = 0;
    
    /**
     * Set up I2C bus to communicate with MAX1704X
     */

    dev.model = MAX17048_9;
    ESP_ERROR_CHECK(max1704x_init_desc(&dev, 0, 18, 19)); // SDA 18, SCL 19 for S3
    ESP_ERROR_CHECK_WITHOUT_ABORT(max1704x_get_config(&dev)); // when I comment this line it works
```

As described in the example above for ESP-32 S3 boards SDA is 18 and SCL is 19, but in this case, on Adafruit with integrated MAX17048 chip I am not sure. 

Does anyone have a clue what could be a problem and did anyone tried to test this on Feather board?

**Comment by UncleRus** [comment]
@shuki25 Please comment

**Comment by shuki25** [comment]
I reviewed the pinout for your board and the pinout for SCL is 4 and SDA is 3.

![The pinout for Adafruit Feather ESP32-S3](https://cdn-learn.adafruit.com/assets/assets/000/110/811/original/adafruit_products_Adafruit_Feather_ESP32-S3_Pinout.png)

**Reply by UncleRus** [reply]
It doesn't matter at all.
Any GPIO can be selected as pins for I2C; the library is specially designed this way.

**Reply by UncleRus** [reply]
Also, everything works if a single line is commented out. (`max1704x_get_config(&dev)`).

**Reply by Max-LAB** [reply]
Hi @UncleRus and @shuki25 thank you for reaching out!

> ```c
> Also, everything works if a single line is commented out. (max1704x_get_config(&dev)).
> ```

Actually, what I wrote up there is not entirely true. 

> ```c
> ESP_ERROR_CHECK_WITHOUT_ABORT(max1704x_get_config(&dev)); // when I comment this line it works
> ```

Sorry for that. It was a long day, and I did not do proper testing. It was a simple two lines of code that got me running the rest of my code, but today, when I tried to get, for example, version `max1704x_get_version(&dev, &version)` I got an error.  

So here is an actual state when using `max1704x.h` library on [Adafruit ESP32 S3 Feather board](https://learn.adafruit.com/adafruit-esp32-s3-feather/power-management)

Truth is for sure that if I use SDA 18 and SCL 19 and try for example simply to get a version board freeze

**Example code with SDA 18, SCL 19**

```c
#include <stdio.h>
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>
#include <max1704x.h>
#include <esp_err.h>
#include <esp_log.h>

static char *TAG = "main";

void test(void *pvParameters)
{

    ESP_LOGI(TAG, "Voltage calculator is starting...");

    esp_err_t r;
    max1704x_t dev = {0 };
    max1704x_config_t config = { 0 };
    max1704x_status_t status = { 0 };
    uint16_t version = 0;
    float voltage = 0;
    float soc_percent = 0;
    float rate_change = 0;
    
    /**
     * Set up I2C bus to communicate with MAX1704X
     */

    dev.model = MAX17048_9;

    ESP_ERROR_CHECK(max1704x_init_desc(&dev, 0, 18, 19));
    ESP_ERROR_CHECK(max1704x_get_version(&dev, &version));
    ESP_LOGI(TAG, "Version: %d\n", version);
    

    while (1)
    {
        ESP_LOGI(TAG, "Get values...");
        r = max1704x_get_voltage(&dev, &voltage);
       
        printf("\n");
        vTaskDelay(pdMS_TO_TICKS(5000));
    }

}

void app_main()
{
    ESP_LOGI(TAG, "Voltage example...");
    /**
     * Initialize I2C bus
     */
    ESP_ERROR_CHECK(i2cdev_init());
    ESP_LOGI(TAG, "Starting test task...");
    xTaskCreate(test, "test", configMINIMAL_STACK_SIZE * 8, NULL, 5, NULL);
}
```

The output that I get using `idf.py -p /dev/cu.usbmodem1413201 monitor`

```bash
➜  default idf.py -p /dev/cu.usbmodem1413201 monitor
Executing action: monitor
Running idf_monitor in directory /Users/max/esp/max1704x/default
Executing "/Users/max/.espressif/python_env/idf5.1_py3.11_env/bin/python /Users/max/esp/esp-idf/tools/idf_monitor.py -p /dev/cu.usbmodem1413201 -b 115200 --toolchain-prefix xtensa-esp32s3-elf- --target esp32s3 --revision 0 /Users/max/esp/max1704x/default/build/example-max1704x.elf -m '/Users/max/.espressif/python_env/idf5.1_py3.11_env/bin/python' '/Users/max/esp/esp-idf/tools/idf.py' '-p' '/dev/cu.usbmodem1413201'"...
--- esp-idf-monitor 1.3.1 on /dev/cu.usbmodem1413201 115200 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
None
Waiting for the device to reconnect.
```


On the other hand, if I use any other SDA, and SCL combination including `SDA=3` and `SCL=4` which is described as pinout for my board, the board will not freeze but I will get this error.

`E (1338) i2cdev: Could not read from device [0x36 at 0]: 263 (ESP_ERR_TIMEOUT)`


**Example code with SDA 3, SCL 4**
```c
// begining of code

void test(void *pvParameters)
{


    dev.model = MAX17048_9;

    ESP_ERROR_CHECK(max1704x_init_desc(&dev, 0, 3, 4));
    ESP_ERROR_CHECK(max1704x_get_version(&dev, &version));
    ESP_LOGI(TAG, "Version: %d\n", version);

// rest of the code    
```

The output that I get using `idf.py -p /dev/cu.usbmodem1413201 monitor`

```sh
ELF file SHA256: 60b4e5dc48937232

Rebooting...
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x3 (RTC_SW_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40375710
0x40375710: esp_restart_noos_dig at /Users/max/esp/esp-idf/components/esp_system/port/esp_system_chip.c:57 (discriminator 1)

SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3818,len:0x16e0
load:0x403c9700,len:0x4
load:0x403c9704,len:0xc00
load:0x403cc700,len:0x2eb0
entry 0x403c9908
I (26) boot: ESP-IDF v5.1.1 2nd stage bootloader
I (26) boot: compile time Dec  3 2023 17:14:12
I (26) boot: Multicore bootloader
I (29) boot: chip revision: v0.1
I (33) boot.esp32s3: Boot SPI Speed : 80MHz
I (38) boot.esp32s3: SPI Mode       : DIO
I (43) boot.esp32s3: SPI Flash Size : 2MB
I (47) boot: Enabling RNG early entropy source...
I (53) boot: Partition Table:
I (56) boot: ## Label            Usage          Type ST Offset   Length
I (64) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (71) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (79) boot:  2 factory          factory app      00 00 00010000 00124f80
I (86) boot: End of partition table
I (90) esp_image: segment 0: paddr=00010020 vaddr=3c020020 size=0b0b8h ( 45240) map
I (107) esp_image: segment 1: paddr=0001b0e0 vaddr=3fc92a00 size=02828h ( 10280) load
I (109) esp_image: segment 2: paddr=0001d910 vaddr=40374000 size=02708h (  9992) load
I (118) esp_image: segment 3: paddr=00020020 vaddr=42000020 size=1a7cch (108492) map
I (144) esp_image: segment 4: paddr=0003a7f4 vaddr=40376708 size=0c2d8h ( 49880) load
I (161) boot: Loaded app from partition at offset 0x10000
I (161) boot: Disabling RNG early entropy source...
I (172) cpu_start: Multicore app
I (173) cpu_start: Pro cpu up.
I (173) cpu_start: Starting app cpu, entry point is 0x40375344
0x40375344: call_start_cpu1 at /Users/max/esp/esp-idf/components/esp_system/port/cpu_start.c:154

I (0) cpu_start: App cpu up.
I (191) cpu_start: Pro cpu start user code
I (191) cpu_start: cpu freq: 160000000 Hz
I (191) cpu_start: Application information:
I (194) cpu_start: Project name:     example-max1704x
I (200) cpu_start: App version:      1
I (204) cpu_start: Compile time:     Dec  3 2023 17:13:44
I (210) cpu_start: ELF file SHA256:  60b4e5dc48937232...
I (216) cpu_start: ESP-IDF:          v5.1.1
I (221) cpu_start: Min chip rev:     v0.0
I (226) cpu_start: Max chip rev:     v0.99
I (231) cpu_start: Chip rev:         v0.1
I (236) heap_init: Initializing. RAM available for dynamic allocation:
I (243) heap_init: At 3FC95AD0 len 00053C40 (335 KiB): DRAM
I (249) heap_init: At 3FCE9710 len 00005724 (21 KiB): STACK/DRAM
I (256) heap_init: At 3FCF0000 len 00008000 (32 KiB): DRAM
I (262) heap_init: At 600FE010 len 00001FD8 (7 KiB): RTCRAM
I (269) spi_flash: detected chip: gd
I (272) spi_flash: flash io: dio
W (276) spi_flash: Detected size(8192k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (290) sleep: Configure to isolate all GPIO pins in sleep state
I (296) sleep: Enable automatic switching of GPIO sleep configuration
I (304) app_start: Starting scheduler on CPU0
I (308) app_start: Starting scheduler on CPU1
I (308) main_task: Started on CPU0
I (318) main_task: Calling app_main()
I (318) main: Voltage example...
I (328) main: Starting test task...
I (328) main: Voltage calculator is starting...
I (338) main_task: Returned from app_main()
E (1338) i2cdev: Could not read from device [0x36 at 0]: 263 (ESP_ERR_TIMEOUT)
ESP_ERROR_CHECK failed: esp_err_t 0x107 (ESP_ERR_TIMEOUT) at 0x42008740
0x42008740: test at /Users/max/esp/max1704x/default/main/main.c:31 (discriminator 1)

file: "./main/main.c" line 31
func: test
expression: max1704x_get_version(&dev, &version)

abort() was called at PC 0x4037aceb on core 0
0x4037aceb: _esp_error_check_failed at /Users/max/esp/esp-idf/components/esp_system/esp_err.c:50



Backtrace: 0x4037588e:0x3fc9d740 0x4037acf5:0x3fc9d760 0x4038084a:0x3fc9d780 0x4037aceb:0x3fc9d7f0 0x42008740:0x3fc9d820 0x4037d3ad:0x3fc9d8a0
0x4037588e: panic_abort at /Users/max/esp/esp-idf/components/esp_system/panic.c:452

0x4037acf5: esp_system_abort at /Users/max/esp/esp-idf/components/esp_system/port/esp_system_chip.c:84

0x4038084a: abort at /Users/max/esp/esp-idf/components/newlib/abort.c:38

0x4037aceb: _esp_error_check_failed at /Users/max/esp/esp-idf/components/esp_system/esp_err.c:50

0x42008740: test at /Users/max/esp/max1704x/default/main/main.c:31 (discriminator 1)

0x4037d3ad: vPortTaskWrapper at /Users/max/esp/esp-idf/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:162
```

**Reply by shuki25** [reply]
Also, I was looking at the Adafruit website, and it appears that previous board versions had an older battery management IC. Can you confirm that your board is equipped with the latest version featuring the MAX1704X chip?

If so, did you run menuconfig? There's a setting in there that you have to specify which chipset you are using.

Component config > MAX1704X Battery Fuel Gauge > Chip type

I reviewed the [schematic of the board](https://learn.adafruit.com/adafruit-esp32-s3-feather/downloads), it's still showing that SDA is 3 and SCL is 4. So if your code says it could not read from device 0x36, it probably means that the device isn't there on the I2C bus line.

I would suggest you to use ESP-IDF example code to scan all devices on the i2c bus line to see what devices you have connected on the board.

[ESP-IDF example code for i2c_tools](https://github.com/espressif/esp-idf/tree/master/examples/peripherals/i2c/i2c_tools)

**Reply by Max-LAB** [reply]
@shuki25 thank you for your reply

> Also, I was looking at the Adafruit website, and it appears that previous board versions had an older battery management IC. Can you confirm that your board is equipped with the latest version featuring the MAX1704X chip?

Actually yes, my board is the one with a `MAX17048 Monitor` label at the back and `7048/1CA6` written on chip on the front like in here [ESP32 Feather Pinouts](https://learn.adafruit.com/assets/118387)

Here is my original picture

![IMG_4120](https://github.com/UncleRus/esp-idf-lib/assets/5176180/332a47e4-5d17-49c1-8dd1-0b9ba68870c2)

> If so, did you run menuconfig? There's a setting in there that you have to specify which chipset you are using.
> Component config > MAX1704X Battery Fuel Gauge > Chip type

Actually yes, this also occurred to me

![Bildschirmfoto 2023-12-05 um 21 10 14](https://github.com/UncleRus/esp-idf-lib/assets/5176180/cfcfa5e1-5ab7-4402-bf63-b448997dbb89)

And here is a snippet from `sdkconfig` file which also confirms that

![Bildschirmfoto 2023-12-05 um 21 12 06](https://github.com/UncleRus/esp-idf-lib/assets/5176180/f7d60f45-93b2-4b89-996a-4feb82b98ece)

Would posting the entire `sdkconfig` file be beneficial?

>I would suggest you to use ESP-IDF example code to scan all devices on the i2c bus line to see what devices you have connected on the board.

Aaand here is the culprit, nothing is found on i2c bus line :(

![Bildschirmfoto 2023-12-05 um 21 41 20](https://github.com/UncleRus/esp-idf-lib/assets/5176180/fc64ef56-4719-4baa-a4e5-8acfd0590a2e)

The Qeustion is how to enable it or wake it up? Do I need a pullup resistor between SDA and +3.3V and also SCL and +3.3V? Nothing is mentioned in here [MAX1708 Battery Monitor](https://learn.adafruit.com/adafruit-esp32-s3-feather/pinouts#max17048-battery-monitor-3137370)

**Reply by Max-LAB** [reply]
@shuki25 

Small progress, on i2c-tools test above wrong SDA and SCL were selected. When I changed them to SDA 3 and SCL 4 scan started but unfortunately, nothing was found. Here is a code and picture

cmd_i2ctools.c
```c
static const char *TAG = "cmd_i2ctools";

#if CONFIG_IDF_TARGET_ESP32S3 || CONFIG_IDF_TARGET_ESP32H2
static gpio_num_t i2c_gpio_sda = 3; #<-- correct SDA
static gpio_num_t i2c_gpio_scl = 4; #<-- correct SCL
#elif CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32C2
static gpio_num_t i2c_gpio_sda = 5;
static gpio_num_t i2c_gpio_scl = 6;
#else
static gpio_num_t i2c_gpio_sda = 18;
static gpio_num_t i2c_gpio_scl = 19;
#endif

static uint32_t i2c_frequency = 100000;
static i2c_port_t i2c_port = I2C_NUM_0;

```
![Bildschirmfoto 2023-12-08 um 23 58 52](https://github.com/UncleRus/esp-idf-lib/assets/5176180/a177f732-16f2-4920-83e0-d83ea467d518)

**Reply by Max-LAB** [reply, marked as answer]
@shuki25 Finally, I managed to get it to work on my [Adafruit ESP32 S3 Feather board](https://learn.adafruit.com/adafruit-esp32-s3-feather/power-management) :) 

In the end problem was the pullup resistors between SDA - 3V and SCL - 3V. Also, the last comment in this thread helped me. [Esp32 IDF release/5.0 running on ESP32S3 Feather I2C problem](https://www.esp32.com/viewtopic.php?t=31449). For this implementation I used 3k resistors. 

![Bildschirmfoto 2023-12-09 um 15 38 49](https://github.com/UncleRus/esp-idf-lib/assets/5176180/3b74925c-0688-46bb-8657-21607d9e0eea)

It is not the best and most beautiful implementation, but for the test it was sufficient.

![IMG_4141](https://github.com/UncleRus/esp-idf-lib/assets/5176180/cb345ba3-5086-44f5-b115-b7b8413088a6)

i2c-tools detected it immediately. SDA 3 - SCl 4

![Bildschirmfoto 2023-12-09 um 15 43 48](https://github.com/UncleRus/esp-idf-lib/assets/5176180/e250d365-c549-4cf5-9e79-a2321c728228)

And here is an output from your library :)

![Bildschirmfoto 2023-12-09 um 15 42 39](https://github.com/UncleRus/esp-idf-lib/assets/5176180/97c015f6-91c8-4dc8-98b0-d35072b09c89)

Thank you both @shuki25 and @UncleRus  again for the quick responses and help you provided.

**Reply by shuki25** [reply]
That's great to hear that you figured it out. Seems like the 10K pull-up resistors was too high for i2c bus line in the original schematic design. 

Read the following article, useful information about this. 

[i2c pull-up resistance calculation article](https://fastbitlab.com/i2c-pull-up-resistance-rise-time-and-bus-capacitance-discussion/)

**Comment by Max-LAB** [comment]
Hi @shuki25, hi @UncleRus I have one additional not directly related question to an upper problem. For my project, I want to determine the entire power consumption of my sensor network. My question is, can I rely on SOC and Voltage parameters received and collected from every node during some period of time? 

Clearly, i can see voltage and SOC percentage drop during some period of time, and plot it inside a graph but can this be a good indicator of how much power my sensor network is using during runtime? 

My test scenario is 5 nodes scattered throughout a closed space. I did 3 measurements with all nodes in the same places, the same code deployed on every one of them, the same antenna orientation, and closely the same battery SOC at the beginning of the measurement. In all three times, I got almost the same values with 0,01750 Voltage and 0,34766 SOC percentage difference at the end.

**Reply by shuki25** [reply]
As far as what I already know from different sources is that the IC chip's operation involves the calculation of the State of Charge (SOC) value over time. Therefore, repeatedly resetting the MAX1704x chip may lead to reduced accuracy, as it continuously monitors charging and battery usage over an extended period.
